### ISSUE TITLE
TypeFilter example is misleading with how ActionFilterAttribute is used

### ISSUE THREAD

**Issue by poke** [aspnet/Docs#9095]
[The documentation on `[TypeFilter]`](https://docs.microsoft.com/en-us/aspnet/core/mvc/controllers/filters?view=aspnetcore-2.1#typefilterattribute) is using a somewhat misleading example:

```c#
[TypeFilter(typeof(AddHeaderAttribute), Arguments = new object[] { "Author", "Steve Smith (@ardalis)" })]
public IActionResult Hi(string name)
{
    return Content($"Hi {name}");
}
```

This suggests that the combination of `[TypeFilter]` and a filter attribute implementation (`AddHeaderAttribute`) is a useful combination. However, `[TypeFilter]` is actually supposed to be used with _any_ kind of filter implementation, and a filter attribute that inherits from `~FilterAttribute` is meant to be used as an attribute directly.

I understand that `AddHeaderAttribute` was probably chosen because it was a filter that was implemented earlier in that topic (so readers know it already). But it was introduced as a [filter attribute](https://docs.microsoft.com/en-us/aspnet/core/mvc/controllers/filters?view=aspnetcore-2.1#built-in-filter-attributes) that can be used directly as an attribute. And that’s the primary reason why you would inherit from `~FilterAttribute`: To implement a filter and a filter factory at the same time that can be used as an attribute.

Since the type filter example is part of the [dependency injection topic](https://docs.microsoft.com/en-us/aspnet/core/mvc/controllers/filters?view=aspnetcore-2.1#dependency-injection), using the `AddHeaderAttribute` filter is actually a bad fit anyway since it does not require any service dependencies.

If the filter implementation _had_ additional service dependencies, it would be a better example for the DI topic. For example this implementation that has an additional service dependency:

```c#
public class AddHeaderWithDependenciesAttribute : ResultFilterAttribute
{
        public AddHeaderWithDependenciesAttribute(string name, string value, ISomeService someService)
        {
            // …
        }

        // …
}
```

This could be added like this, and it would work just fine since that’s what the `TypeFilter` attribute does:

```c#
[TypeFilter(typeof(AddHeaderWithDependenciesAttribute), Arguments = new object[] { "Author", "Steve Smith (@ardalis)" })]
public IActionResult Hi(string name) { … }
```

However, since it’s a filter attribute, inheriting from `ResultFilterAttribute`, I would also expect to be able to use it as a filter directly but that won’t work because the constructor will also require the service dependency:

```c#
[AddHeaderWithDependenciesAttribute("foo", "bar", !?!?!)]
public IActionResult Hi(string name) { … }
```

---

So, long story short, I would appreciate if the example for `TypeFilter` would actually use a filter implementation (not a filter attribute!) that used both passed and injected arguments. For example:

```c#
[TypeFilter(typeof(AddHeaderWithDependenciesFilter), Arguments = new object[] { "Example value" })]
public IActionResult Example()
{
    return Content("Example");
}

public class AddHeaderWithDependenciesFilter : IResultFilter
{
    private readonly string _value;
    private readonly ILogger<AddHeaderWithDependenciesFilter> _logger;
    public AddHeaderWithDependenciesFilter(string value, ILogger<AddHeaderWithDependenciesFilter> logger)
    {
        _value = value;
        _logger = logger;
    }

    public void OnResultExecuted(ResultExecutedContext context)
    { }
    public void OnResultExecuting(ResultExecutingContext context)
    { }

    // …
}
```

**Comment by Rick-Anderson** [comment]
@dougbu please review and provide new sample code (or suggest another reviewer).

**Comment by Rick-Anderson** [comment]
@dougbu please review  (or suggest another reviewer).

**Comment by dougbu** [comment]
@Rick-Anderson I agree the current example is a bit confusing for the reasons listed.

The sample code @poke provided looks fine. If there's a PR to review, I can do it. But, i doubt I'll have much to say.
